### ISSUE TITLE
implicit declaration of function ‘OPENSSL_cleanup breaks a libressl based Gentoo ejabberd server

### ISSUE THREAD

**Issue by toralf** [processone/fast_tls#11]
At a stable hardened Gentoo with libressl I do get for 1.0.7

 QA Notice: Package triggers severe warnings which indicate that it
            may exhibit random runtime failures.
 c_src/fast_tls_drv.c:350:5: warning: implicit declaration of function ‘OPENSSL_cleanup’ [-Wimplicit-function-declaration]
 c_src/fast_tls_drv.c:470:8: warning: implicit declaration of function ‘DH_set0_pqg’ [-Wimplicit-function-declaration]

and indeed this correlates to https://bugs.gentoo.org/show_bug.cgi?id=600932

few more info:
=ERROR REPORT==== 27-Nov-2016::17:26:32 ===
failed to load TLS driver: undefined symbol: OPENSSL_cleanup
*failed*
in function fast_tls:'-load_nif_test/0-fun-0-'/0 (src/fast_tls.erl, line 432)
**error:{assertEqual,[{module,fast_tls},
              {line,432},
              {expression,"load_driver ( )"},
              {expected,ok},
              {value,{error,{open_error,-10}}}]}
  output:<<"">>

module 'p1_sha'
  p1_sha: load_nif_test...ok
  p1_sha: sha1_test...ok
  p1_sha: sha224_test...ok
  p1_sha: sha256_test...ok
  p1_sha: sha384_test...ok
  p1_sha: sha512_test...ok
  p1_sha: to_hexlist_test...ok
  [done in 0.021 s]

**Comment by prefiks** [comment]
So it looks like module compiles code paths that are specific for openssl 1.1+ but then is used with lib that still uses interface for older version of openssl.

It's possible that you have headers from openssl1.1 somewhere and those are used instead for that version of lib? Conditional compilation is based on OPENSSL_VERSION_NUMBER that is in file opensslv.h, could you check if you have duplicates of that in your system?

**Comment by toralf** [comment]
Hi,

at this server I do use libressl instead of openssl and do have this file only once:



mr-fox ~ # cat /usr/include/openssl/opensslv.h
/* $OpenBSD: opensslv.h,v 1.36 2016/06/30 11:10:29 bcook Exp $ */
#ifndef HEADER_OPENSSLV_H
#define HEADER_OPENSSLV_H

/* These will change with each release of LibreSSL-portable */
#define LIBRESSL_VERSION_NUMBER 0x2050000fL
#define LIBRESSL_VERSION_TEXT   "LibreSSL 2.5.0"

/* These will never change */
#define OPENSSL_VERSION_NUMBER  0x20000000L
#define OPENSSL_VERSION_TEXT    LIBRESSL_VERSION_TEXT
#define OPENSSL_VERSION_PTEXT   " part of " OPENSSL_VERSION_TEXT

#define SHLIB_VERSION_HISTORY ""
#define SHLIB_VERSION_NUMBER "1.0.0"

#endif /* HEADER_OPENSSLV_H */

**Comment by prefiks** [comment]
This should be fixed by commit 09cf2def53d97a0567d685618bb5b3d2547f4b59

**Comment by toralf** [comment]
I applied that to 1.0.7 but still get : 
```
=ERROR REPORT==== 28-Nov-2016::11:00:15 ===
fxailed to load TLS driver: undefined symbol: OPENSSL_cleanup
module 'p1_sha'
  p1_sha: load_nif_test...[0.013 s] ok
  p1_sha: sha1_test...ok
  p1_sha: sha224_test...ok
  p1_sha: sha256_test...ok
  p1_sha: sha384_test...ok
  p1_sha: sha512_test...ok
  p1_sha: to_hexlist_test...ok
  [done in 0.034 s]
=======================================================
  Failed: 1.  Skipped: 0.  Passed: 7.
Cover analysis: /var/tmp/portage/dev-erlang/fast_tls-1.0.7/work/fast_tls-1.0.7/.eunit/index.html
Coverdata export: /var/tmp/portage/dev-erlang/fast_tls-1.0.7/work/fast_tls-1.0.7/.eunit/cover.coverdata
ERROR: One or more eunit tests failed.
ERROR: eunit failed while processing /var/tmp/portage/dev-erlang/fast_tls-1.0.7/work/fast_tls-1.0.7: reb
```

**Comment by prefiks** [comment]
Could you please check current master? I think it should work now, please also call 'make' before 'make test' as i think the later one doesn't compile code before running tests.

**Comment by toralf** [comment]
same problem even with HEAD (and yes, "make" is made before "test" ;) )
1.0.3 BTW is fine and passes all tests

**Comment by prefiks** [comment]
I really don't know what's going then, i did build libressl from source and i can compile and run test fine with it here. Could you see if 'make clean; make;l make test' will build correct version for you?

**Comment by toralf** [comment]
[libressl 2,5,0 installed] 

tfoerste@mr-fox ~ $  git clone https://github.com/processone/fast_tls.git
Cloning into 'fast_tls'...
remote: Counting objects: 481, done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 481 (delta 9), reused 0 (delta 0), pack-reused 461
Receiving objects: 100% (481/481), 178.79 KiB | 0 bytes/s, done.
Resolving deltas: 100% (300/300), done.
Checking connectivity... done.

tfoerste@mr-fox ~ $ cd fast_tls/

tfoerste@mr-fox ~/fast_tls $ make clean; make; make test
rebar clean
 ==> fast_tls (clean)
rebar get-deps compile
==> fast_tls (get-deps)
Pulling p1_utils from {git,"git://github.com/processone/p1_utils",
                           {tag,"1.0.6"}}
Cloning into 'p1_utils'...
==> p1_utils (get-deps)
==> p1_utils (compile)
Compiled src/p1_nif_utils.erl
Compiled src/p1_http.erl
Compiled src/p1_fsm.erl
Compiled src/p1_time_compat.erl
Compiled src/treap.erl
Compiled src/p1_server.erl
Compiled src/p1_edoc_layout.erl
==> fast_tls (compile)
Compiled src/fast_tls_sup.erl
Compiled src/p1_sha.erl
Compiled src/fast_tls_app.erl
Compiled src/fast_tls.erl
Compiling c_src/fast_tls_drv.c
Compiling c_src/p1_sha.c
rebar skip_deps=true eunit
==> fast_tls (eunit)
Compiled src/fast_tls_sup.erl
Compiled src/p1_sha.erl
Compiled src/fast_tls.erl
Compiled src/fast_tls_app.erl
fast_tls: load_nif_test (module 'fast_tls')...
=ERROR REPORT==== 28-Nov-2016::15:32:19 ===
failed to load TLS driver: undefined symbol: OPENSSL_cleanup
*failed*
in function fast_tls:'-load_nif_test/0-fun-0-'/0 (src/fast_tls.erl, line 436)
**error:{assertEqual,[{module,fast_tls},
              {line,436},
              {expression,"load_driver ( )"},
              {expected,ok},
              {value,{error,{open_error,-10}}}]}
  output:<<"">>

=======================================================
  Failed: 1.  Skipped: 0.  Passed: 7.
Cover analysis: /home/tfoerste/fast_tls/.eunit/index.html
Coverdata export: /home/tfoerste/fast_tls/.eunit/cover.coverdata
ERROR: One or more eunit tests failed.
ERROR: eunit failed while processing /home/tfoerste/fast_tls: rebar_abort
Makefile:10: recipe for target 'test' failed
make: *** [test] Error 1

**Comment by prefiks** [comment]
Could you please do this for me: execute "rebar clean; rebar compile -v|grep fast_tls_drv.c"
This should give you something like: 

cmd: cc -c  -DHAS_ERTS_EXIT -g -O2 -Wall -g -Wall -fPIC -MMD  -I"/usr/lib/erlang/lib/erl_interface-3.9.1/include" -I"/usr/lib/erlang/erts-8.1.1/include"   c_src/fast_tls_drv.c -o c_src/fast_tls_drv.o

could you then take that part after cmd:, and replace part after -o with "-o my_processed_file.c -E" to get something like this:
cc -c  -DHAS_ERTS_EXIT -g -O2 -Wall -g -Wall -fPIC -MMD  -I"/usr/lib/erlang/lib/erl_interface-3.9.1/include" -I"/usr/lib/erlang/erts-8.1.1/include"   c_src/fast_tls_drv.c -o my_processed_file.c -E

this should create my_processed_file.c, could you attach this file if you get it?

**Comment by toralf** [comment]
https://zwiebeltoralf.de/pub/my_processed_file.c

**Comment by prefiks** [comment]
Thank you.

So this file shows that it uses code paths that are should be compatible with libressl/openssl < 1.1 (for example there is no mention of OPENSSL_cleanup that "make test" complain on your machine).

I wonder if by chance it doesn't pick version of this library that is installed on your system, and is picked by erlang before path from source, if you execute "rebar shell" in source dir and then after it shows prompt type this:

p1_nif_utils:get_so_path(fast_tls, [fast_tls], "fast_tls_drv").

do you get path that point to where you have that repo cloned or to system path?

**Comment by prefiks** [comment]
Could you also please check if this 
nm -D priv/lib/fast_tls_drv.so |grep OPENSSL_cleanup
shows something?

**Comment by toralf** [comment]
tfoerste@mr-fox ~/fast_tls $ rebar shell
==> fast_tls (shell)
Erlang/OTP 19 [erts-8.1] [source] [64-bit] [async-threads:10]

Eshell V8.1  (abort with ^G)
1> p1_nif_utils:get_so_path(fast_tls, [fast_tls], "fast_tls_drv").
"/usr/lib64/erlang/lib/fast_tls-1.0.7/priv/lib/fast_tls_drv"

tfoerste@mr-fox ~/fast_tls $ ^C
tfoerste@mr-fox ~/fast_tls $ nm -D priv/lib/fast_tls_drv.so |grep OPENSSL_cleanup

**Comment by toralf** [comment]
pffr, if I apply the mentioned commit 09cf2 onto 1.0.7 and *install* fast-tls-1.0.7 here at the stable Gentoo server, then starting ejabberd yields into a different crash now:  

2016-11-28 17:12:51.380 [error] <0.131.0> failed to load TLS driver: undefined symbol: DH_set0_pqg
2016-11-28 17:12:51.381 [error] <0.131.0> CRASH REPORT Process <0.131.0> with 0 neighbours exited with reason: {open_error,-10} in gen_server:init_it/6 line 344
2016-11-28 17:12:51.385 [error] <0.130.0> Supervisor fast_tls_sup had child fast_tls started with fast_tls:start_link() at undefined exit with reason {open_error,-10} in context start_error
2016-11-28 17:12:51.385 [critical] <0.62.0>@ejabberd:exit_or_halt:133 failed to start application 'fast_tls': {error,
                                         {{shutdown,
                                           {failed_to_start_child,fast_tls,
                                            {open_error,-10}}},
                                          {fast_tls_app,start,[normal,[]]}}}
2016-11-28 17:12:51.386 [error] <0.128.0> CRASH REPORT Process <0.128.0> with 0 neighbours exited with reason: {{shutdown,{failed_to_start_child,fast_tls,{open_error,-10}}},{fast_tls_app,start,[normal,[]]}} in application_master:init/4 line 134


so it seems, that
1. the cloned repo preferres to use installed libs rather then the one in the cloned repo, or ?
2. there's still another issue
or ?

**Comment by prefiks** [comment]
Latest master contains fix that should make it always use locally build .so instead of picking system one.

**Comment by prefiks** [comment]
You would need to apply more commit that just that one(as i just copy pasted check for libressl from one place in others, where some of them should have been negated) - you would need to also apply two after that.

**Comment by toralf** [comment]
indeed, 1.0.7-9-g6e08c8b passes test now fine
and ejabberd starts now fine after applying "git diff 1.0.7..onto fast_tls package
but crashes now with something completely different (maybe ejabberd is not ready with latest fast_tls)  : 

mr-fox jabber # cat error.log 
2016-11-28 17:36:45.872 [error] <0.61.0> CRASH REPORT Process <0.61.0> with 0 neighbours exited with reason: {aborted,{bad_type,{ejabberd_commands,gen_html_doc_for_commands,[documentation],"Generates html documentation for ejabberd_commands",[],0,ejabberd_commands_doc,generate_html_output,[{file,binary},{regexp,binary},{examples,binary}],restricted,[],{res,rescode},["Path to file where generated documentation should be stored","Regexp matching names of commands or modules that will be included inside generated document","Comma separated list of languages (choosen from java, perl, xmlrpc, json)that ..."],...}}} in application_master:init/4 line 134
